### PR TITLE
Add package upgrade rules

### PR DESCRIPTION
**Pull request by mmoskal**
Also improvements to compile switches and variants

### CODE REVIEW

File: localtypings/pxtarget.d.ts [hunk 1]
```diff
@@ -37,6 +37,10 @@ declare namespace pxt {
         bannedRepos?: string[];
         allowUnapproved?: boolean;
         preferredRepos?: string[]; // list of company/project(#tag) of packages to show by default in search
+        // format:
+        // "acme-corp/pxt-widget": "min:v0.1.2" - auto-upgrade to that version
+        // "acme-corp/pxt-widget": "dv:foo,bar" - add "disablesVariant": ["foo", "bar"] to pxt.json
+        upgrades?: pxt.Map<string>;
```
Comment (pelikhan) [inline localtypings/pxtarget.d.ts:43]: do we want to have a mini format like or just have a JSON structure to describe those ranages? what happens if you want to combine min:v0.1.2,dv:foo,bar.

For the target, we have something like this:         
```
patches?: pxt.Map<UpgradePolicy[]>; // semver range -> upgrade policies
```

File: pxtcompiler/emitter/hexfile.ts [hunk 2]
```diff
@@ -1088,13 +1088,16 @@ _stored_program: .hex ${res}
     }
 
     function assembleAndPatch(src: string, bin: Binary, opts: CompileOptions, cres: CompileResult) {
-        if (opts.extinfo.disabledDeps) {
+        const dummy = opts.extinfo.disabledDeps
```
Comment (pelikhan) [inline pxtcompiler/emitter/hexfile.ts:1091]: extinfo could be undefined if the C++ build failed?

File: pxtlib/github.ts [hunk 3]
```diff
@@ -557,7 +557,16 @@ namespace pxt.github {
             return Promise.resolve<CachedPackage>(undefined);
         }
 
-        return db.loadPackageAsync(p.fullName, p.tag);
+        return db.loadPackageAsync(p.fullName, p.tag)
+            .then(cached => {
+                const dv = upgradedDisablesVariants(config, repoWithTag)
+                if (dv) {
+                    const cfg = JSON.parse(cached.files[pxt.CONFIG_NAME]) as pxt.PackageConfig
```
Comment (pelikhan) [inline pxtlib/github.ts:564]: use ``Package.parseAndValidConfig``

File: localtypings/pxtarget.d.ts (hunk 1 above)
Comment (mmoskal) [reply localtypings/pxtarget.d.ts:43]: I'm not a great fan of using JSON as a DSL... in this case there's also no need to combine - if you enforce a minimum version, you also make sure it has the right properties in its pxt.json

File: pxtcompiler/emitter/hexfile.ts (hunk 2 above)
Comment (mmoskal) [reply pxtcompiler/emitter/hexfile.ts:1091]: I don't believe we get here when the build failed; there was an unguarded reference down below for a while

File: pxtlib/github.ts (hunk 3 above)
Comment (mmoskal) [reply pxtlib/github.ts:564]: thanks!

File: localtypings/pxtarget.d.ts (hunk 1 above)
Comment (mmoskal) [reply localtypings/pxtarget.d.ts:43]: In fact I think the package min version should probably move to UpgradePolicy. Currently, we show the wrong version number in explorer after upgrade.

Comment (mmoskal) [reply localtypings/pxtarget.d.ts:43]: OK, I'm mistaken. I'll just apply it in the place where UpgradePolicies are applied.

Comment (mmoskal) [reply localtypings/pxtarget.d.ts:43]: OK done. Package rules are now still in targetconfig.json so they can be updated out of band. The package refs are updated in pxt.json and persisted, and disabledVariants are done on the fly.

File: pxtlib/github.ts [hunk 4]
```diff
@@ -557,7 +557,18 @@ namespace pxt.github {
...
 
-        return db.loadPackageAsync(p.fullName, p.tag);
+        return db.loadPackageAsync(p.fullName, p.tag)
+            .then(cached => {
+                const dv = upgradedDisablesVariants(config, repoWithTag)
+                if (dv) {
+                    const cfg = Package.parseAndValidConfig(cached.files[pxt.CONFIG_NAME])
+                    if (cfg) {
+                        cfg.disablesVariants = dv
+                        cached.files[pxt.CONFIG_NAME] = JSON.stringify(cfg, null, 4)
```
Comment (pelikhan) [inline pxtlib/github.ts:567]: We also have an api to strongly the pxt.json . I think stringingConfig or something. It makes sure that fields are in same order etc

File: pxtlib/github.ts [hunk 5]
```diff
@@ -557,7 +557,18 @@ namespace pxt.github {
...
         }
 
-        return db.loadPackageAsync(p.fullName, p.tag);
+        return db.loadPackageAsync(p.fullName, p.tag)
+            .then(cached => {
+                const dv = upgradedDisablesVariants(config, repoWithTag)
+                if (dv) {
+                    const cfg = Package.parseAndValidConfig(cached.files[pxt.CONFIG_NAME])
+                    if (cfg) {
+                        cfg.disablesVariants = dv
```
Comment (pelikhan) [inline pxtlib/github.ts:566]: Add a debug log entry?

File: pxtlib/package.ts [hunk 6]
```diff
@@ -318,6 +318,24 @@ namespace pxt {
             return regex.test(ts);
         }
 
+        private upgradePackagesAsync() {
+            return pxt.packagesConfigAsync()
+                .then(config => {
+                    let numfixes = 0
+                    U.iterMap(this.config.dependencies, (pkg, ver) => {
+                        if (pxt.github.isGithubId(ver)) {
+                            const upgraded = pxt.github.upgradedPackageReference(config, ver)
```
Comment (pelikhan) [inline pxtlib/package.ts:327]: probably good to first normalize the GitHub id

Comment (pelikhan) [reply pxtlib/package.ts:327]:
```
                            ver = pxt.github.normalizeRepoId(ver);
```

Comment (mmoskal) [reply pxtlib/package.ts:327]: we parse it on the other side, so it's probably OK

**Comment by mmoskal** [comment]
good to go as soon as build passes

**Pull request merged**